homeassistant/components/casatunes/media_player.py

```diff
@@ -0,0 +1,273 @@
+
+    for source in sources:
+        try:
+            player_devices.append(CasaTunesPlayer(source))
+        except TypeError:
+            _LOGGER.error("Unable to create a player")
+            raise PlatformNotReady
+
+    async_add_entities(zone_devices)
+    async_add_entities(player_devices)
+
+
+class CasaTunesPlayer(MediaPlayerDevice):
+    """Representation of a CasaTunes player."""
+
+    def __init__(self, player):
+        """Initialize the CasaTunes player."""
+        self._player = player
+
+    async def async_update(self):
+        """Get the latest details."""
+        await self._player.update()
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_MUSIC
+
+    @property
+    def media_content_id(self):
+        """Return the content ID of current playing media."""
+        return self._player.content_id
+
+    @property
+    def media_title(self):
+        """Title of current playing media."""
+        return self._player.title
+
+    @property
+    def media_artist(self):
+        """Artist of current playing media, music track only."""
+        return self._player.artist
+
+    @property
+    def media_album_name(self):
+        """Album name of current playing media, music track only."""
+        return self._player.album
+
+    @property
+    def media_image_url(self):
+        """Image url of current playing media."""
+        return self._player.image_url
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._player.name
+
+    @property
+    def media_duration(self):
+        """Return the duration of current playing media in seconds."""
+        return self._player.duration
+
+    @property
+    def media_track(self):
+        """Return the track number of current media (Music track only)."""
+        return self._player.track
+
+    @property
+    def shuffle(self):
+        """Boolean if shuffle is enabled."""
+        return self._player.shuffle
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        status = self._player.status
+
+        if status == 0:
+            return STATE_IDLE
+        elif status == 1:
+            return STATE_PAUSED
+        elif status == 2:
+            return STATE_PLAYING
+        else:
+            return STATE_ON
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORT_CT_PLAYER
+
+    @property
+    def media_position(self):
+        """Position of current playing media in seconds."""
+        return self._player.position
+
+    @property
+    def media_position_updated_at(self):
+        """When was the position of the current playing media valid."""
+        return self._player.position_updated_at
+
+    async def async_media_seek(self, position):
+        """Send seek command."""
+        await self._player.seek(position)
```

Comment:
Does this trigger an update so media position reflects the correct placement?

Reply:
Yes, it does!

Reply:
Unfortunately no, but I think I found workaround. Still need to do a bit of testing.

Reply:
Given this integration uses polling, you could:
1. Update state here if the new data would be available in fresh poll: `self.async_schedule_update_ha_state(True)`.  `True` indicates to call `async_update` before updating the state machine.
1. Operate optimistically by setting the new position and last update and then call `self.async_schedule_update_ha_state()`

You may need something similar in the other commands if you're finding the state isn't updating until the next poll.

Reply:
Thanks for the hint! I think, I'll have to spend a bit more time with it as the implementation of CasaTunes "seeking" is not very reliable.